// @nx-extend/core/src/utils/exec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execCommand = exports.ChildProcess = exports.ShellString = void 0;
const shell = require("shelljs");
const shelljs_1 = require("shelljs");
Object.defineProperty(exports, "ShellString", { enumerable: true, get: function () { return shelljs_1.ShellString; } });
const child_process_1 = require("child_process");
Object.defineProperty(exports, "ChildProcess", { enumerable: true, get: function () { return child_process_1.ChildProcess; } });
const execCommand = (command, options = {
    asString: false,
    asJSON: false
}) => {
    if (!options.silent) {
        console.log('\nRunning: ', command);
    }
    const result = shell.exec(command, options);
    if (options.asJSON) {
        return JSON.parse(result.toString());
    }
    if (options.asString) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        return result.toString();
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    return {
        success: result['code'] === 0,
        output: result.stdout
    };
};
exports.execCommand = execCommand;
